crud/stats/module.lua

```diff
@@ -0,0 +1,234 @@
+--
+function stats.enable(opts)
+    checks({ driver = '?string' })
+
+    opts = opts or {}
+    opts.driver = opts.driver or 'local'
+
+    StatsError:assert(drivers[opts.driver] ~= nil, 'Unsupported driver: %s', opts.driver)
+
+    if stats_registry == drivers[opts.driver] then
+        return true
+    end
+
+    -- Disable old driver registry, if another one was requested.
+    stats.disable()
+
+    stats_registry = drivers[opts.driver]
+    stats_registry.init()
+
+    return true
+end
+
+--- Destroys statistics registry and disable callbacks
+--
+-- @function disable
+--
+-- @treturn[1] boolean Returns true.
+--
+function stats.disable()
+    if stats.is_enabled() then
+        stats_registry.destroy()
+        stats_registry = nil
+    end
+
+    return true
+end
+
+--- Get statistics on CRUD operations
+--
+-- @function get
+--
+-- @treturn table Statistics on CRUD operations.
+--  If statistics disabled, returns {}.
+--
+function stats.get()
+    if stats_registry ~= nil then
+        return stats_registry.get()
+    end
+
+    return {}
+end
+
+--- Wrap CRUD operation call to collect statistics
+--
+-- @function wrap
+--
+-- @tparam function func
+--  Function to wrap. Errors are caught and thrown again.
+--
+-- @tparam string op_label
+--  Label of registry collectos.
+--  Use `require('crud.common.const').OP` to pick one.
+--
+-- @return First two arguments of wrapped function output.
+--
+function stats.wrap(func, op_label)
+    dev_checks('function', 'string')
+    return function(...)
+        if not stats.is_enabled() then
+            return func(...)
+        end
+
+        local start_time = clock.monotonic()
+        local status, res, err = pcall(func, ...)
+        local finish_time = clock.monotonic()
+
+        local success = status and (err == nil)
+        local latency = finish_time - start_time
+        stats_registry.observe(op_label, success, latency)
+
+        if status == false then
+            error(res)
+        end
+
+        return res, err
+    end
+end
+
+--- Wrap pairs operation call to collect statistics
+--
+-- @function wrap_pairs
+--
+-- @tparam function func
+--  Pairs function to wrap. Errors are caught and thrown again.
+--
+-- @tparam string op_label
+--  Label of registry collectos.
+--  Use `require('crud.common.const').OP` to pick one.
+--
+-- @return Returns pairs generator.
+--
+function stats.wrap_pairs(func, op_label)
+    dev_checks('function', 'string')
+    return function(...)
+        if not stats.is_enabled() then
+            return func(...)
+        end
+
+        local start_time = clock.monotonic()
+        local status, gen, param, state = pcall(func, ...)
+        local finish_time = clock.monotonic()
+
+        local success = status
+        local latency = finish_time - start_time
+        stats_registry.observe(op_label, success, latency)
+
+        if status == false then
+            error(gen)
+        end
+
+        return gen, param, state
+    end
+end
```

Comment:
Looks the same as stats.wrap. Could it be generalized?

Probably you could rewrite it with tail call (see box.atomic implementation)

Reply:
Tail call is really something I was interested in, but couldn't find am example, thanks. But there is another issue. For general function call, getting a second argument is treated as an error and we increase error collectors, but for pairs `gen, param, state` is a default.

Reply:
Reworked